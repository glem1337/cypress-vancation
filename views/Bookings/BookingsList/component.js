import React from 'react';
import { Button, Input, Pagination } from 'antd';
import classNames from 'classnames';
import { FormattedMessage } from 'react-intl';

import { PAGINATION_SIZE } from 'constants/booking';

import BookingCard from '../BookingCard';

import useContainer from './hook';

const BookingsList = () => {
  const {
    containerRef,
    isContainerScrolled,
    isSearchExpanded,
    toggleSearchExpandCondition,
    setFiltersMenuVisibility,
    camperInquiries,
    camperInquiriesPage,
    camperInquiriesTotal,
    onPaginationChange,
  } = useContainer();

  return (
    <div
      className={classNames('chat-list', isContainerScrolled && 'chat-list--scrolled')}
      ref={containerRef}
    >
      <div className="chat-list__header">
        <div className="d-flex align-items-center mr-auto">
          <p className="text-headline mr-8">
            <FormattedMessage id="shared.inbox" />
          </p>
          <span className="main-account-header__item-counter">2</span>
        </div>
        <div>
          <Button
            size="small"
            icon={<i className="icon icon-search" />}
            className="chat-list-filter-btn"
            onClick={toggleSearchExpandCondition}
          />
          <div
            className={classNames('chat-list__search-input', {
              'chat-list__search-input--open': isSearchExpanded,
            })}
          >
            <Input prefix={<i className="icon icon-search" />} allowClear />
            <Button
              onClick={toggleSearchExpandCondition}
              size="large"
              className="chat-list__search-input-btn"
              icon={<i className="icon icon-cross in-gray-500" />}
            />
          </div>
          {/* TODO:
           if filter is selected, add type="secondary" to Button
            and new class "in-blue-1000 to icon"
          */}
          <Button
            size="small"
            icon={<i className="icon icon-filter" />}
            className="chat-list-filter-btn"
            onClick={setFiltersMenuVisibility(true)}
          />
        </div>
      </div>
      <div>
        {camperInquiries.map((item, index) => (
          <BookingCard
            key={item.id}
            isFirst={index === 0}
            id={item.id}
          />
        ))}
        {camperInquiriesTotal > PAGINATION_SIZE && (
          <div className="d-flex flex-column align-items-flex-end mt-24 mr-20 mt-md-40">
            <Pagination
              current={camperInquiriesPage}
              total={camperInquiriesTotal}
              pageSize={PAGINATION_SIZE}
              showSizeChanger={false}
              onChange={onPaginationChange}
            />
          </div>
        )}
      </div>
    </div>
  );
};

export default BookingsList;
