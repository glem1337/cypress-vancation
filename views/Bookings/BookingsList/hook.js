import { useState, useRef, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { FETCH_USER_TYPE } from 'constants/booking';
import { setFiltersVisibility, fetchCamperInquiries } from 'state/concepts/booking/actions';
import {
  camperInquiriesSelector,
  camperInquiresTotalSelector,
  camperInquiresPageSelector,
} from 'state/concepts/booking/selectors';

function useContainer() {
  const [isSearchExpanded, setIsSearchExpanded] = useState(false);
  const [isContainerScrolled, setIsContainerScrolled] = useState(false);

  const containerRef = useRef();

  const isContainerScrolledRef = useRef(false);

  const dispatch = useDispatch();

  const camperInquiries = useSelector(camperInquiriesSelector);
  const camperInquiriesPage = useSelector(camperInquiresPageSelector);
  const camperInquiriesTotal = useSelector(camperInquiresTotalSelector);

  /**
   * Toggle search expand condition.
   */
  const toggleSearchExpandCondition = () => {
    setIsSearchExpanded(prev => !prev);
  };

  /**
   * Set filters menu visibility.
   */
  const setFiltersMenuVisibility = (isVisible) => () => {
    dispatch(setFiltersVisibility(isVisible));
  };

  /**
   * On pagination change
   */
  const onPaginationChange = (page) => {
    dispatch(fetchCamperInquiries({
      page,
      userType: FETCH_USER_TYPE.RENTER,
    }));
  };

  /**
   * Scroll handler.
   */
  const scrollHandler = () => {
    if (!containerRef.current) {
      return false;
    }

    const isScrolled = containerRef.current.scrollTop !== 0;

    if (isContainerScrolledRef.current !== isScrolled) {
      setIsContainerScrolled(isScrolled);
    }

    return true;
  };

  /**
   * Set scroll listeners.
   */
  const setScrollListeners = () => {
    if (!containerRef.current?.addEventListener) {
      return false;
    }

    containerRef.current.addEventListener('scroll', scrollHandler);

    return true;
  };

  /**
   * Remove scroll listeners.
   */
  const removeScrollListeners = () => {
    if (!containerRef.current?.removeEventListener) {
      return false;
    }

    containerRef.current.removeEventListener('scroll', scrollHandler);

    return true;
  };

  /**
   * Mounting
   */
  useEffect(() => {
    setScrollListeners();

    return removeScrollListeners;
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return {
    containerRef,
    isContainerScrolled,
    isSearchExpanded,
    toggleSearchExpandCondition,
    setFiltersMenuVisibility,
    setScrollListeners,
    removeScrollListeners,
    scrollHandler,
    camperInquiries,
    camperInquiriesPage,
    camperInquiriesTotal,
    onPaginationChange,
  };
}

export default useContainer;
