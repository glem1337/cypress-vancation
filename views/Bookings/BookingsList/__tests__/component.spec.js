import React from 'react';
import { shallow } from 'enzyme';

import BookingsList from '../component';

const mockedHookData = {
  containerRef: {},
  isContainerScrolled: true,
  isSearchExpanded: {},
  toggleSearchExpandCondition: jest.fn(),
  setFiltersMenuVisibility: jest.fn(),
  camperInquiries: [{ id: 1 }, { id: 2 }],
  camperInquiriesPage: 1,
  camperInquiriesTotal: 22,
  onPaginationChange: jest.fn(),
};
jest.mock('../hook', () => jest.fn(() => mockedHookData));

describe('BookingsList component tests', () => {
  const component = shallow(<BookingsList />);

  it('matches snapshot', () => {
    expect(component).toMatchSnapshot();
  });
});
