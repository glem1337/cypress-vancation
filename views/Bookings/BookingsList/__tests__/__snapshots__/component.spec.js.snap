// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingsList component tests matches snapshot 1`] = `
<div
  className="chat-list chat-list--scrolled"
>
  <div
    className="chat-list__header"
  >
    <div
      className="d-flex align-items-center mr-auto"
    >
      <p
        className="text-headline mr-8"
      >
        <Memo(FormattedMessage)
          id="shared.inbox"
        />
      </p>
      <span
        className="main-account-header__item-counter"
      >
        2
      </span>
    </div>
    <div>
      <Button
        className="chat-list-filter-btn"
        icon={
          <i
            className="icon icon-search"
          />
        }
        onClick={[MockFunction]}
        size="small"
      />
      <div
        className="chat-list__search-input chat-list__search-input--open"
      >
        <Input
          allowClear={true}
          prefix={
            <i
              className="icon icon-search"
            />
          }
          type="text"
        />
        <Button
          className="chat-list__search-input-btn"
          icon={
            <i
              className="icon icon-cross in-gray-500"
            />
          }
          onClick={[MockFunction]}
          size="large"
        />
      </div>
      <Button
        className="chat-list-filter-btn"
        icon={
          <i
            className="icon icon-filter"
          />
        }
        size="small"
      />
    </div>
  </div>
  <div>
    <BookingCard
      id={1}
      isFirst={true}
      key="1"
    />
    <BookingCard
      id={2}
      isFirst={false}
      key="2"
    />
    <div
      className="d-flex flex-column align-items-flex-end mt-24 mr-20 mt-md-40"
    >
      <Pagination
        current={1}
        onChange={[MockFunction]}
        pageSize={20}
        showSizeChanger={false}
        total={22}
      />
    </div>
  </div>
</div>
`;
